src/Main.hs

```diff
@@ -189,33 +173,35 @@ splitBehaviour store (Transition name contract iface@(Interface _ decls) iffs' c
       storageBounds = snd $ getStorageBounds env
   iff <- checkIffs env (iffs' <> calldataBounds <> storageBounds)
   postcondition <- mapM (checkBool env) (fromMaybe [] maybePost)
-  flatten iff postcondition [] cases
+  flatten iff postcondition cases
   where
     env = mkEnv contract store decls
 
     -- translate wildcards into negation of other cases
-    normalize :: [Case Expr b] -> [Case Expr b]
-    normalize =
-      snd
-        . mapAccumL
-          ( \a b -> case b of
-              Leaf pn WildExp p -> (a, Leaf pn (ENot nowhere (andRaw a)) p)
-              Branch pn WildExp p -> (a, Branch pn (ENot nowhere (andRaw a)) p)
-              e@(Leaf _ c _) -> (c : a, e)
-              e@(Branch _ c _) -> (c : a, e)
-          )
-          []
-
-    -- flatten case tree
-    flatten :: [Exp Bool] -> [Exp Bool] -> [Exp Bool] -> Case Expr Post -> Err [Claim]
-    flatten iff postc pathcond (Leaf _ cond post) = do
-      c <- checkBool env cond
-      (p, maybeReturn, contracts) <- checkPost env contract post
-      return $ splitCase name False contract iface (c : pathcond) iff maybeReturn p postc contracts
-    flatten iff postc pathcond (Branch _ cond cs) = do
-      c <- checkBool env cond
-      leaves <- mapM (flatten iff postc (c : pathcond)) (normalize cs)
-      return $ join leaves
+    normalize :: [Case] -> [Case]
+    normalize cases = snd $ mapAccumL checkCase (BoolLit False) cases
+
+    checkCase :: Expr -> Case -> (Expr, Case)
+    checkCase acc (Case p WildExp post) =
+      (error "wildcard not last case",
+        Case p (ENot nowhere acc) post)
+    checkCase acc (Case p e post) = (EOr nowhere e acc, Case p e post)
+
+
+    -- flatten case list
+    flatten :: [Exp Bool] -> [Exp Bool] -> Cases -> Err [Claim]
+    flatten iff postc (Direct post) = do
+      (p, maybeReturn) <- checkPost env contract post
+      return $ splitCase name False contract iface (LitBool True) iff maybeReturn p postc
+    flatten iff postc (Branches branches) = do
+      let branches' = normalize branches
+      cases <- flip mapM branches $ \(Case pn cond post) -> do
+        if' <- checkBool env cond
+        (post', ret) <- checkPost env contract post
+        return (if', post', ret)
+
+      pure . join $ ((\(ifcond, ret, store) ->
```

Comment:
shouldn't this be `(ifcond, stateUpdates, ret)` or similar?

Reply:
yes, I screwed up these argnames